File: frontend/app/lib/audio.ts
import { useAuthenticatedFetch } from "./authFetch";

export interface AudioUploadRequest {
  storyId: number;
  lineNumber: number;
  label: string;
  fileName: string;
}

export interface AudioUploadResponse {
  uploadUrl: string;
  filePath: string;
  fileBucket: string;
}

export interface AudioConfirmRequest {
  storyId: number;
  lineNumber: number;
  filePath: string;
  fileBucket: string;
  label: string;
}

export class AudioUploadError extends Error {
  constructor(
    message: string,
    public step: "request" | "upload" | "confirm",
  ) {
    super(message);
    this.name = "AudioUploadError";
  }
}

export function createAudioUploader() {
  const authenticatedFetch = useAuthenticatedFetch();

  return async function uploadAudioFile(
    file: File,
    storyId: number,
    lineNumber: number,
    label: string = "complete",
  ): Promise<void> {
    // Step 1: Request upload URL
    const uploadRequest: AudioUploadRequest = {
      storyId,
      lineNumber,
      label,
      fileName: file.name,
    };

    let uploadResponse: AudioUploadResponse;
    try {
      const response = await authenticatedFetch(
        "/api/admin/stories/audio/upload",
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(uploadRequest),
        },
      );

      if (!response.ok) {
        throw new Error(`Failed to request upload URL: ${response.status}`);
      }

      uploadResponse = await response.json();
    } catch (error) {
      throw new AudioUploadError(
        error instanceof Error ? error.message : "Failed to request upload URL",
        "request",
      );
    }

    // Step 2: Upload file to signed URL
    try {
      const uploadFileResponse = await fetch(uploadResponse.uploadUrl, {
        method: "PUT",
        body: file,
        headers: {
          "Content-Type": file.type,
        },
      });

      if (!uploadFileResponse.ok) {
        throw new Error(`File upload failed: ${uploadFileResponse.status}`);
      }
    } catch (error) {
      throw new AudioUploadError(
        error instanceof Error ? error.message : "File upload failed",
        "upload",
      );
    }

    // Step 3: Confirm upload
    const confirmRequest: AudioConfirmRequest = {
      storyId,
      lineNumber,
      filePath: uploadResponse.filePath,
      fileBucket: uploadResponse.fileBucket,
      label,
    };

    try {
      const confirmResponse = await authenticatedFetch(
        "/api/admin/stories/audio/confirm",
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(confirmRequest),
        },
      );

      if (!confirmResponse.ok) {
        throw new Error(`Failed to confirm upload: ${confirmResponse.status}`);
      }
    } catch (error) {
      throw new AudioUploadError(
        error instanceof Error ? error.message : "Failed to confirm upload",
        "confirm",
      );
    }
  };
}

export function createAudioDeleter() {
  const authenticatedFetch = useAuthenticatedFetch();

  return async function deleteLineAudio(
    storyId: number,
    lineNumber: number,
  ): Promise<void> {
    try {
      const response = await authenticatedFetch(
        "/api/admin/stories/audio/delete",
        {
          method: "DELETE",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ storyId, lineNumber }),
        },
      );

      if (!response.ok) {
        throw new Error(`Failed to delete audio: ${response.status}`);
      }
    } catch (error) {
      throw new AudioUploadError(
        error instanceof Error ? error.message : "Failed to delete audio",
        "request",
      );
    }
  };
}

File: src/admin/stories/audio.go
package stories

import (
	"encoding/json"
	"net/http"
	"strconv"
	"strings"
	"time"

	"glossias/src/auth"
	"glossias/src/pkg/models"
)

type AudioUploadRequest struct {
	StoryID    int    `json:"storyId"`
	LineNumber int    `json:"lineNumber"`
	Label      string `json:"label"`
	FileName   string `json:"fileName"`
}

type AudioUploadResponse struct {
	UploadURL  string `json:"uploadUrl"`
	FilePath   string `json:"filePath"`
	FileBucket string `json:"fileBucket"`
}

const (
	bucket = "audio-files"
)

func (h *Handler) audioUploadHandler(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case "POST":
		h.requestAudioUploadURL(w, r)
	default:
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
	}
}

func (h *Handler) requestAudioUploadURL(w http.ResponseWriter, r *http.Request) {
	var req AudioUploadRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	// Admin authentication check
	userID, ok := auth.GetUserIDWithOk(r)
	if !ok || !models.CanUserEditStory(r.Context(), userID, int32(req.StoryID)) {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	// Validate request data
	if req.StoryID <= 0 || req.LineNumber <= 0 || req.Label == "" || req.FileName == "" {
		http.Error(w, "Invalid Request: Data is outside bounds. Ensure all data is present and try again.", http.StatusBadRequest)
		return
	}

	// Check if story exists
	exists, err := models.StoryExists(r.Context(), int32(req.StoryID))
	if err != nil {
		h.log.Error("Failed to check story existence", "error", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}
	if !exists {
		http.Error(w, "Story not found", http.StatusNotFound)
		return
	}

	// Check if line exists
	lineExists, err := models.LineExists(r.Context(), req.StoryID, req.LineNumber)
	if err != nil {
		h.log.Error("Failed to check line existence", "error", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}
	if !lineExists {
		http.Error(w, "Line not found", http.StatusNotFound)
		return
	}

	// Check if audio file already exists for this line and label
	existingAudioFiles, err := models.GetLineAudioFiles(r.Context(), req.StoryID, req.LineNumber)
	if err != nil {
		h.log.Error("Failed to check existing audio files", "error", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}
	for _, audioFile := range existingAudioFiles {
		if audioFile.Label == req.Label {
			http.Error(w, "Audio file with this label already exists for this line", http.StatusConflict)
			return
		}
	}

	// Generate file path: stories/{storyID}/line_{lineNumber}_{label}_{filename}
	timestamp := time.Now().Unix()
	// Sanitize filename to prevent path traversal
	sanitizedFilename := strings.ReplaceAll(req.FileName, "/", "")
	sanitizedFilename = strings.ReplaceAll(sanitizedFilename, "\\", "")
	sanitizedFilename = strings.ReplaceAll(sanitizedFilename, "..", "")
	filePath := "stories/" + strconv.Itoa(req.StoryID) + "/line_" +
		strconv.Itoa(req.LineNumber) + "_" + req.Label + "_" +
		strconv.FormatInt(timestamp, 10) + "_" + sanitizedFilename

	// Generate signed upload URL
	signedURL, err := models.GenerateSignedUploadURL(r.Context(), bucket, filePath)
	if err != nil {
		h.log.Error("Failed to generate signed upload URL", "error", err)
		http.Error(w, "Failed to generate upload URL", http.StatusInternalServerError)
		return
	}

	response := AudioUploadResponse{
		UploadURL:  signedURL,
		FilePath:   filePath,
		FileBucket: bucket,
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func (h *Handler) audioDeleteHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != "DELETE" {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	type DeleteAudioRequest struct {
		StoryID    int `json:"storyId"`
		LineNumber int `json:"lineNumber"`
	}

	var req DeleteAudioRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	// Admin authentication check
	userID, ok := auth.GetUserIDWithOk(r)
	if !ok || !models.CanUserEditStory(r.Context(), userID, int32(req.StoryID)) {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	// Validate request data
	if req.StoryID <= 0 || req.LineNumber <= 0 {
		http.Error(w, "Invalid request: storyId and lineNumber must be positive", http.StatusBadRequest)
		return
	}

	// Delete all audio files for the line
	err := models.DeleteLineAudioFiles(r.Context(), req.StoryID, req.LineNumber)
	if err != nil {
		h.log.Error("Failed to delete line audio files", "error", err)
		http.Error(w, "Failed to delete audio files", http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

func (h *Handler) confirmAudioUploadHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != "POST" {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	type ConfirmUploadRequest struct {
		StoryID    int    `json:"storyId"`
		LineNumber int    `json:"lineNumber"`
		FilePath   string `json:"filePath"`
		FileBucket string `json:"fileBucket"`
		Label      string `json:"label"`
	}

	var req ConfirmUploadRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	// Admin authentication check for confirm step
	userID, ok := auth.GetUserIDWithOk(r)
	if !ok || !models.CanUserEditStory(r.Context(), userID, int32(req.StoryID)) {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	// Verify file path matches expected pattern to prevent path manipulation
	expectedPrefix := "stories/" + strconv.Itoa(req.StoryID) + "/line_" + strconv.Itoa(req.LineNumber) + "_" + req.Label + "_"
	if !strings.HasPrefix(req.FilePath, expectedPrefix) {
		http.Error(w, "Invalid file path", http.StatusBadRequest)
		return
	}

	// Create audio file record in database
	audioFile, err := models.CreateAudioFile(r.Context(), req.StoryID, req.LineNumber,
		req.FilePath, req.FileBucket, req.Label)
	if err != nil {
		if err == models.ErrAudioFileExists {
			http.Error(w, "An audio file with this label already exists for this line", http.StatusBadRequest)
			return
		}
		h.log.Error("Failed to create audio file record", "error", err)
		http.Error(w, "Failed to create audio file record", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(audioFile)
}
